pkg/addonmanager/controllers/agentdeploy/default_sync.go

```diff
@@ -0,0 +1,48 @@
+package agentdeploy
+
+import (
+	"context"
+	"fmt"
+
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/constants"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	"open-cluster-management.io/addon-framework/pkg/basecontroller/factory"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+type defaultSyncer struct {
+	controller *addonDeployController
+	agentAddon agent.AgentAddon
+}
+
+func (s *defaultSyncer) sync(ctx context.Context,
+	syncCtx factory.SyncContext,
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (*addonapiv1alpha1.ManagedClusterAddOn, error) {
+	installMode := constants.InstallModeDefault
+	deployWorkNamespace := addon.Namespace
+	deployWorkName := constants.DeployWorkName(addon.Name)
+	deployWork, _, err := s.controller.buildManifestWorks(ctx, s.agentAddon, installMode, deployWorkNamespace, cluster, addon)
+	if err != nil {
+		return addon, err
+	}
+
+	// deployWork is nil only in the case that there is no object from addon manifests. need to delete the deployWork.
+	if deployWork == nil {
+		err = deleteWork(ctx, s.controller.workClient, deployWorkNamespace, deployWorkName)
+		if err != nil {
+			return addon, fmt.Errorf("failed to delete work %s/%s when got no object", deployWorkNamespace, deployWorkName)
+		}
+		s.controller.cache.removeCache(deployWorkNamespace, deployWorkName)
+		return addon, nil
+	}
+
+	if !addon.DeletionTimestamp.IsZero() {
```

Comment:
Is it better to move these before `s.controller.buildManifestWorks`?

Reply:
and I am thinking should we also need to deleteWork(...) when `!addon.DeletionTimestamp.IsZero()`?

Reply:
yes. done.